Fix PEP 0 name parsing

AA-Turner opened the pull request:
This fixes name parsing in PEP 0 for generation of the indicies, for example `Ernest W. Durbin III`, `The Python core team and community`, `Eric N. Vander Weele`. 

This fix is also included in PR #1385, but as per @merwok's suggestion in #2 I'm creating this new pull request for the single issue.

Thanks,
Adam

merwok commented on pep0/pep.py:
```diff
@@ -62,34 +62,38 @@ class Author(object):
+
+        name_dict = self._parse_name(name)
+        self.suffix = name_dict.get("suffix")
+        if name_dict.get("name"):
```
```suggestion
        if "name" in name_dict:
```

(avoid retrieving and discarding the value)

merwok commented on pep0/pep.py:
```diff
@@ -62,34 +62,38 @@ class Author(object):
-            self.last_first += " (%s)" % (self.nick,)
-        else:
-            self.nick = self.last
+                raise ValueError(f"unknown van Rossum {self}!")
```
```suggestion
                raise ValueError(f"unknown van Rossum {self!r}!")
```
to keep any space, quotes, etc.

AA-Turner replied:
Having just tested the Author class doesn't print nicely normally or with the repr flag, so will change to printing the inputted name

merwok commented on pep0/pep.py:
```diff
@@ -109,28 +113,64 @@ def sort_by(self):
+                surname = " ".join(name_parts[-2:])
+                name.update(forename=forename, surname=surname)
+
+            # handles double surnames after a middle initial (e.g. N. Vander Weele)
```
This is tough, a name like `R. David Murray` should give `Murray` as a surname for the PEP index.

That said, a name written in Japanese English convention like `INADA Naoki` should return `INADA` as surname.
In short, it’s not generally possible to parse world names into US categories of «first», «middle», «last»

I guess we have to accept imperfect for now, add special cases when we notice problems (what PEP will add the first `Name MacName Sr` :slightly_smiling_face:), and maybe someday rework the system to have the proper way: metadata (or some author index dict) should include full name and short name for PEP 0.

AA-Turner replied:
Luckily (?!) in PEP 458 `R. David Murray` is listed without the full stop in his first initial, so the name is still parsed correctly. I don't think there's a good solution for `INADA Naoki` beyond adding a special-case exception

> In short, it’s not generally possible to parse world names into US categories of «first», «middle», «last»

Or UK, Australia, Canada etc :stuck_out_tongue_winking_eye:. But I get your point. I'm reminded of [this post](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/) about names - hopefully #&#8203;40 doesn't apply to us...

I think that your last suggestion having some sort of lookup table is probably the best solution, as in all the PEPs there are still only a relativley small number of authors (248) - it's quite late here so will add that feature tommorow. It also keeps special cases etc. out of the code to keep it from becoming knobbly.

merwok replied:
It would be OK if this PEP fixed the most egregious cases (`III`) without covering 100% of possibilities :slightly_smiling_face:

AA-Turner replied:
Gives me something to do! Latest commit adds such a metadata lookup and therefore simplifies the name parsing code. 

This should make it so that names can be correctly entered into AUTHORS.csv and PEP 0 will reflect this. I've also identified some duplicate entries (e.g. `P.J. Eby` &  `Phillip J. Eby`, `Greg` and `Gregory` Ewing, `Jim J. Jewett` & Jim Jewett`, `Martin v. Löwis` & `Martin von Löwis`). Is it acceptable to modify PEP headers to canonicalise these names?

merwok replied:
I think I would add multiple entries to the data file rather than editing historical documents.

merwok replied:
The latest data file (exceptions rather than full mapping) doesn’t de-duplicate these entries, should it?

AA-Turner replied:
The data file is checked first (in __init__), so unsure where duplicates would propogate from? 

Always good to be preventative but not sure I understand this one, sorry!

merwok replied:
Maybe my comment doesn’t make sense!
I don’t have a clear picture of the current behaviour of the code, so I wondered if the change from full data file to exceptions data file did preserve the feature you added of normalizing the duplicate names.

AA-Turner replied:
Aha! I forgot to add that back in, you're right - have done so now (only adding the less used variant and mapping it to the 'cannonical' variant, to keep the file smaller)

merwok commented on AUTHORS.csv:
```diff
@@ -0,0 +1,249 @@
+"Full Name";                            "Surname First"; 						"Name Reference"
```
If this file was really comma-separated rather than semicolon-plus-space, I think github would display it as a sortable, searchable table.

AA-Turner replied:
Good shout - semicolons were for speed as CSV didn't like commas within the text, but I'll try escaping them. Likewise spacing was to make it easier to edit the pure text file, but if github will render intelligently it would be worth removing these I think.

merwok replied:
Seems OK to avoid extra spaces, people can use all kinds of plugins in text editors or IDEs to edit a CSV file, or even a graphical spreadsheet program as long as they are careful when saving.

AA-Turner replied:
Changed, and github does indeed render it nicely!

AA-Turner commented:
Is there anything outstanding to do on this one @merwok? / What would next steps be?

merwok commented:
Is there a preview up somewhere?  If not, I’ll have a look on my laptop.

AA-Turner commented:
> 
> 
> Is there a preview up somewhere? If not, I’ll have a look on my laptop.

I haven't set one up, sorry

vstinner commented:
Hum, I understand that the purpose of this PEP is to fix:

> "I | 8100 | January 2019 steering council election | Smith, III"

"Ernest W. Durbin III" becomes "III".

How is the CSV file generated? How is it supposed to be maintained?

I would prefer to only store "exceptions" in this file: if a name is "<first name> <surname>" (two words), we can pick the second word as the surname.

I'm not sure why my name "Victor Stinner" is rendered as "Stinner, Victor". Why not simply copying the name unchanged? Is it a convention?
https://www.python.org/dev/peps/pep-0000/#authors-owners

Also I'm not sure if it's a good idea to provide a long list of email addresses.

AA-Turner commented:
> Hum, I understand that the purpose of this PEP is to fix:
> 
> > "I | 8100 | January 2019 steering council election | Smith, III"
> 
> How is the CSV file generated? How is it supposed to be maintained?
> 
> I would prefer to only store "exceptions" in this file: if a name is `<first name> <surname>` (two words), we can pick the second word as the surname.

I think Eric also gave examples of surnames from a [non-anglicised tradition ](https://en.wikipedia.org/wiki/Personal_name#Name_order)(e.g. `INADA Naoki`) which this process should also fix. However I think I like the proposal of a sort of 'authors-exception' file better, which would also be easier to maintain as you note. I'll work on this now.

> I'm not sure why my name "Victor Stinner" is rendered as "Stinner, Victor". Why not simply copying the name unchanged? Is it a convention?
> [python.org/dev/peps/pep-0000/#authors-owners](https://www.python.org/dev/peps/pep-0000/#authors-owners)

I'd be happy to render the name unchanged, but don't want to make unilateral changes for obvious reasons! Rendering the name unchanged would make all of these workarounds unnecessary, so would be easier from a maintenance perspective.

Having done a bit of looking, the `Last, First` format was in @benjaminp's original PEP0 generator from 12 years ago:
https://github.com/python/peps/blob/1fbe18ef4578cfbe7996b53da9417738e15c4e66/pep0/output.py#L177

This seems to have originated from @warsaw in commit b7ac9d0 (Aug 2000). I might suggest that if the Authors/Owners key were to be changed, the names by each PEP should be updated to a similar format.

> Also I'm not sure if it's a good idea to provide a long list of email addresses.

True!

AA-Turner commented:
> I would prefer to only store "exceptions" in this file: if a name is `<first name> <surname>` (two words), we can pick the second word as the surname.
> 

Implemented in latest commit. I haven't removed the long list of names/emails (under heading Authors/Owners), but can do this. I wonder if it should be in a new PR though, to limit scope. Happy to do this if you'd like!

merwok commented on pep0/pep.py:
```diff
@@ -60,36 +60,48 @@ class Author(object):
             if self.first == "Guido":
                 self.nick = "GvR"
             elif self.first == "Just":
                 self.nick = "JvR"
```
I wonder if these special cases could be handled by the CSV file too.

AA-Turner replied:
Done!

merwok commented on AUTHORS.csv:
```diff
@@ -1,3 +1,6 @@
+Inada Naoki, "Inada, Naoki", Inada
+Guido van Rossum, "van Rossum, Guido (GvR)", GvR
+Just van Rossum, "van Rossum, Just (JvR)", JvR
+The Python core team and community, The Python core team and community, The Python core team and community
```
Hah good catch!  If you need a short reference for this last entry, you could use `python-dev`.

AA-Turner replied:
Thanks!

merwok commented:
Hi!  I will finalize this in the coming days, then turn to the Sphinx PRs!

AA-Turner commented:
Thanks! Seems @hugovk is also doing a bunch of pep-infra PRs, so if mine get approval will look to ensure no conflict.

hugovk commented:
I don't think there will be much overlap, if any, but feel free to ping me if you've any questions!

ewdurbin commented:
@merwok Is there a list of outstanding concerns to help finalize this PR? I'd enthusiastically work on them.

it looks like per Zen this ended up going the explicit route for overrides, is it just cleaning up the attempt at automatic handling?

AA-Turner commented:
Re what this PR does, in comparison to #1385 this is pretty single-issue! 

I added an explicit overrides file, and cleaned up a bit of the automatic handling.

merwok commented:
I only wanted to test the branch locally to check the output, but I’ve had no time so far.

@ewdurbin please feel free to take over to approve and merge!

AA-Turner commented:
closing
